### PR TITLE
[Test] [Bug] Trigger the overpruned dependency bug

### PR DESCRIPTION
**Pull request by effectfully**
The added test fails with

```
test
  Plugin
    Data
      9.6
        families
          stakingCredential: FAIL
            Exception: Error: Error from the PIR compiler:
                   Error from the PLC compiler:
                   Free type variable at  Ann {annInline = MayInline, annSrcSpans = { no-src-span }} :  Credential
```

The bug is in the dead code eliminator. To be precise, it's this part that it doesn't handle correctly despite promising to in `PlutusIR.Analysis.Dependencies`:

```
{- Note [Dependencies for datatype bindings, and pruning them]
At face value, all the names introduced by datatype bindings should depend on each other.
Given our meaning of "A depends on B", since we cannot remove any part of the datatype
binding without removing the whole thing, they all depend on each other

However, there are some circumstances in which we *can* prune datatype bindings.

In particular, if the datatype is only used at the type-level (i.e. all the term-level parts
(constructors and destructor) are dead), then we are free to completely replace the binding
with one for a trivial type with the same kind.

This is because there are *no* term-level effects, and types are erased in the end, so
in this case rest of the datatype binding really is superfluous.

But how do we represent this in the dependency graph? We still need to have proper dependencies
so that we don't make the wrong decisions wrt transitively used values, e.g.

let U :: * = ...
let datatype T = T1 | T2 U
in T1

Here we need to not delete U, even though T2 is "dead"!

The solution is to focus on the meaning of "dependency": with the pruning that we can do, we *can*
remove all the term level bits en masse, but only en-mass. So we need to make *them* into a clique,
so that this is visible to the dependency analysis.
-}
```

I don't know what went wrong there.

### CODE REVIEW

**Comment by effectfully** [comment]
@michaelpj any idea?

**Comment by michaelpj** [comment]
What happens if you get rid of the bit where it tries to prune dead constructors?

**Comment by effectfully** [comment]
> What happens if you get rid of the bit where it tries to prune dead constructors?

The test becomes green, see 229b6903412086ed05b468a65e751fb59dd2fb08.

**Comment by michaelpj** [comment]
Since you have a small test case I would dump the output from just after the bad dead code pass and see what's wrong. It *looks* like a simple case we should handle properly...

**Comment by michaelpj** [comment]
From the error it seems like we get rid of `Credential` entirely, which seems odd.

**Comment by effectfully** [comment]
> Since you have a small test case I would dump the output from just after the bad dead code pass and see what's wrong. It looks like a simple case we should handle properly...

Yes, I've done exactly that when investigating and you're right that this is what's happening:

> From the error it seems like we get rid of `Credential` entirely, which seems odd.

**Comment by michaelpj** [comment]
I don't see why from just looking at it, some debugging is going to be required I think!

**Comment by effectfully** [comment]
> I don't see why from just looking at it, some debugging is going to be required I think!

Wanna do some debugging on a weekend? :slightly_smiling_face:

File: plutus-benchmark/script-contexts/test/9.6/checkScriptContextEqualityOverhead-20.budget.golden [hunk 1]
```diff
@@ -1,2 +1,2 @@
-({cpu: 34304100
-| mem: 214500})
\ No newline at end of file
+({cpu: 37152100
+| mem: 232300})
```
Comment (effectfully) [inline plutus-benchmark/script-contexts/test/9.6/checkScriptContextEqualityOverhead-20.budget.golden:2]: @zliu41 is this large enough difference to keep the bug in the dead code eliminator that causes necessary constructors to get eliminated?

File: plutus-core/plutus-ir/src/PlutusIR/Analysis/Dependencies.hs [hunk 2]
```diff
@@ -172,23 +172,10 @@ bindingDeps b = case b of
...
-    -- term-level parts depend on each other later, but it's good practice and will be
-    -- useful if we ever stop doing that.
-    destrDeps <-
-      G.overlays
-        <$> (withCurrent destr $ traverse (typeDeps . _varDeclType) constrs)
-    let tus = fmap (view PLC.theUnique) (destr : fmap _varDeclName constrs)
-    -- See Note [Dependencies for datatype bindings, and pruning them]
-    let nonDatatypeClique = G.clique (fmap Variable tus)
-    pure $ G.overlays $ [vDeps] ++ tvDeps ++ cstrDeps ++ [destrDeps] ++ [nonDatatypeClique]
+    let tyus = fmap (view PLC.theUnique) $ _tyVarDeclName d : fmap _tyVarDeclName tvs
```
Comment (zliu41) [inline plutus-core/plutus-ir/src/PlutusIR/Analysis/Dependencies.hs:175]: I'm not quite sure what this does; worth some comment. There should also be some explanation or reference to the bug here.

File: plutus-benchmark/script-contexts/test/9.6/checkScriptContextEqualityOverhead-20.budget.golden (hunk 1 above)
Comment (zliu41) [reply plutus-benchmark/script-contexts/test/9.6/checkScriptContextEqualityOverhead-20.budget.golden:2]: 10% is bad but only a small number of golden files are affected - I'm fine either way. What's your preference?

File: plutus-core/plutus-ir/src/PlutusIR/Analysis/Dependencies.hs (hunk 2 above)
Comment (effectfully) [reply plutus-core/plutus-ir/src/PlutusIR/Analysis/Dependencies.hs:175]: I have no idea either, that's Michael code and as you can see above Michael isn't particularly keen on dealing with it :)

File: plutus-benchmark/script-contexts/test/9.6/checkScriptContextEqualityOverhead-20.budget.golden (hunk 1 above)
Comment (effectfully) [reply plutus-benchmark/script-contexts/test/9.6/checkScriptContextEqualityOverhead-20.budget.golden:2]: I guess let's keep the bug in then, I think it's quite low priority compared to other things and I don't feel like studying the DCE algo at the moment (or ever haha). I'll reflect this problem in an issue.

Comment (zliu41) [reply plutus-benchmark/script-contexts/test/9.6/checkScriptContextEqualityOverhead-20.budget.golden:2]: Sounds good.

**Comment by effectfully** [comment]
I've added an entry to #5967, so that we don't forget about this bug, and let's close this PR for now.

**Comment by zliu41** [comment]
Fixed in https://github.com/IntersectMBO/plutus/pull/6806

**Pull request closed without merging**